### PR TITLE
Add presigned url multi part upload APIs

### PR DESCRIPTION
**Pull request by Ram81**
### Description

This PR adds required APIs for supporting multipart uploads using Presigned URLs. We need this feature inorder to upload huge file in chunks and use the APIs to show progress bar. We are using approach mentioned [here](https://stackoverflow.com/questions/63370330/show-progress-bar-while-uploading-to-s3-using-presigned-url). CLI [PR](https://github.com/Cloud-CV/evalai-cli/pull/291)

### Approach

The workflow for multipart upload using boto3 API:

1. Generate multipart upload id using `create_multipart_upload`
2. Generate `num_chunks` presigned_urls for `upload_id` returned by `create_multipart_upload` API
3. Upload chunks to the new presigned urls and store `ETag` and `PartNumber`
4. Complete upload by calling `complete_multipart_upload` with all the parts with `{'Etag', 'PartNumber'}`

We've created 2 APIs:
1. `get_submission_file_presigned_url` `get_annotation_file_presigned_url`: This API creates a multipart file upload for give `num_file_chunks` and returns the required presigned urls for uploading each part.
2. `submission_file_upload_complete` `annotation_file_upload_complete`: After uploading each file chunk using presigned URL we need to complete multipart upload in order to create the file on S3. This API takes `ETag` and `PartNumber` for each uploaded file chunk and makes a `complete_multipart_upload` API call to boto3.

### CODE REVIEW

File: apps/challenges/urls.py [hunk 1]
```diff
@@ -201,4 +201,9 @@
         views.get_all_challenge_templates,
         name="get_all_challenge_templates",
     ),
+    url(
```
Comment (RishabhJain2018) [inline apps/challenges/urls.py:204]: @Ram81 Is it necessary to have a new API? Can't we use the previous API with the number of chunks parameter?

**Review by RishabhJain2018 — CHANGES_REQUESTED** [review]
Please fix the Travis build as well.

File: apps/challenges/urls.py (hunk 1 above)
Comment (Ram81) [reply apps/challenges/urls.py:204]: @RishabhJain2018 the new API is to complete the chunked upload. The flow is 
1. Generate multipart upload id using `create_multipart_upload`
2. Generate `num_chunks` presigned_urls for `upload_id` returned by `create_multipart_upload` API
3. Upload chunks to the new presigned urls and store `ETag` and `PartNumber`
4. Complete upload by calling `complete_multipart_upload` with all the parts with `{'Etag', 'PartNumber'}`

**Comment by Ram81** [comment]
@RishabhJain2018 This PR is ready for review. I have tested the feature, it works fine

File: apps/challenges/urls.py (hunk 1 above)
Comment (RishabhJain2018) [reply apps/challenges/urls.py:204]: I see..this makes sense. Thanks for the detailed explanation.

File: apps/challenges/utils.py [hunk 2]
```diff
@@ -159,6 +159,88 @@ def generate_presigned_url(file_key_on_s3, challenge_pk):
...
+            )
+            presigned_urls.append({
+                "partNumber": part_number,
+                "url": presigned_url
+            })
+        response_data = {
+            "presigned_urls": presigned_urls,
+            "upload_id": upload_id
+        }
+        return response_data
```
Comment (RishabhJain2018) [inline apps/challenges/utils.py:204]: Can we please use the `return` statement at last only once.

File: apps/challenges/utils.py [hunk 3]
```diff
@@ -159,6 +159,88 @@ def generate_presigned_url(file_key_on_s3, challenge_pk):
...
+    except Exception as e:
+        logger.exception(e)
+        response_data = {"error": "Could not fetch presigned urls."}
+        return response_data
```
Comment (RishabhJain2018) [inline apps/challenges/utils.py:208]: Can we please use the `return` statement at last only once.

File: apps/challenges/utils.py [hunk 4]
```diff
@@ -159,6 +159,88 @@ def generate_presigned_url(file_key_on_s3, challenge_pk):
...
+        s3 = get_boto3_client("s3", aws_keys)
+        response = s3.complete_multipart_upload(
+            Bucket=aws_keys["AWS_STORAGE_BUCKET_NAME"],
+            Key=file_key_on_s3,
+            MultipartUpload={
+                "Parts": parts
+            },
+            UploadId=upload_id
+        )
+        return response
```
Comment (RishabhJain2018) [inline apps/challenges/utils.py:237]: Same as above.

File: apps/challenges/utils.py [hunk 5]
```diff
@@ -159,6 +159,88 @@ def generate_presigned_url(file_key_on_s3, challenge_pk):
...
+    except Exception as e:
+        logger.exception(e)
+        response_data = {"error": "Could not fetch presigned urls."}
+        return response_data
```
Comment (RishabhJain2018) [inline apps/challenges/utils.py:241]: Same as above.

File: apps/challenges/views.py [hunk 6]
```diff
@@ -2680,12 +2681,17 @@ def get_annotation_file_presigned_url(request, challenge_phase_pk):
...
+    if request.data.get("num_file_chunks"):
+        num_file_chunks = int(request.data["num_file_chunks"])
+
     file_ext = os.path.splitext(request.data["file_name"])[-1]
     random_file_name = uuid.uuid4()
 
     if challenge_phase.test_annotation:
         file_key_on_s3 = "{}/{}".format(
-            settings.MEDIAFILES_LOCATION, challenge_phase.test_annotation.name
+            "media", challenge_phase.test_annotation.name
```
Comment (RishabhJain2018) [inline apps/challenges/views.py:2694]: Please don't hardcode the folder names here.

File: apps/challenges/views.py [hunk 7]
```diff
@@ -2710,16 +2716,84 @@ def get_annotation_file_presigned_url(request, challenge_phase_pk):
...
+    """
+    if settings.DEBUG or settings.TEST:
+        response_data = {
+            "error": "Sorry, this feature is not available in development or test environment."
+        }
+        return Response(response_data)
+    # Check if the challenge phase exists or not
+    try:
+        challenge_phase = get_challenge_phase_model(challenge_phase_pk)
+    except ChallengePhase.DoesNotExist:
```
Comment (RishabhJain2018) [inline apps/challenges/views.py:2754]: There is no need to use `try/except` with the `get_challenge_phase_model` function as it automatically handles that case. Please see its definition.

File: apps/challenges/views.py [hunk 8]
```diff
@@ -2710,16 +2716,84 @@ def get_annotation_file_presigned_url(request, challenge_phase_pk):
...
+    if not is_user_a_host_of_challenge(
+        request.user, challenge_phase.challenge.pk
+    ):
+        response_data = {
+            "error": "Sorry, you are not authorized for uploading an annotation file."
+        }
+        return Response(response_data, status=status.HTTP_400_BAD_REQUEST)
+
+    if request.data.get("parts") is None:
+        response_data = {"error": "Uploaded file Parts metadata is missing"}
```
Comment (RishabhJain2018) [inline apps/challenges/views.py:2767]: `Uploaded file Parts metadata is missing` -> `Uploaded file parts metadata is missing!`

File: apps/challenges/views.py [hunk 9]
```diff
@@ -2710,16 +2716,84 @@ def get_annotation_file_presigned_url(request, challenge_phase_pk):
...
+        return Response(response_data, status=status.HTTP_400_BAD_REQUEST)
+
+    if request.data.get("upload_id") is None:
+        response_data = {"error": "Uploaded file UploadId is missing"}
```
Comment (RishabhJain2018) [inline apps/challenges/views.py:2771]: `"Uploaded file UploadId is missing"` -> `"Uploaded file upload Id is missing"`

File: apps/challenges/views.py [hunk 10]
```diff
@@ -2710,16 +2716,84 @@ def get_annotation_file_presigned_url(request, challenge_phase_pk):
...
+        settings.MEDIAFILES_LOCATION, challenge_phase.test_annotation.name
+    )
+
+    try:
+        response = complete_s3_multipart_file_upload(
+            file_parts, upload_id, file_key_on_s3, challenge_phase.challenge.pk
+        )
+        if response.get("error"):
+            response_data = response
+            return Response(response_data, status=status.HTTP_400_BAD_REQUEST)
```
Comment (RishabhJain2018) [inline apps/challenges/views.py:2786]: Same as above for `return` statement.

File: apps/challenges/views.py [hunk 11]
```diff
@@ -2710,16 +2716,84 @@ def get_annotation_file_presigned_url(request, challenge_phase_pk):
...
+        response_data = {
+            "upload_id": upload_id,
+            "challenge_phase_pk": challenge_phase.pk,
+        }
+        return Response(response_data, status=status.HTTP_201_CREATED)
+    except Submission.DoesNotExist:
```
Comment (RishabhJain2018) [inline apps/challenges/views.py:2792]: Why are we checking here for `Submission`? Am I missing something here?

File: apps/challenges/views.py [hunk 12]
```diff
@@ -2710,16 +2716,84 @@ def get_annotation_file_presigned_url(request, challenge_phase_pk):
...
+        response_data = {"error": "Submission does not exist"}
```
Comment (RishabhJain2018) [inline apps/challenges/views.py:2793]: Same as above.

File: apps/jobs/views.py [hunk 13]
```diff
@@ -2301,26 +2302,152 @@ def get_submission_file_presigned_url(request, challenge_phase_pk):
...
+    Returns:
+         Response Object -- An object containing the presignd url and submission id, or an error message if some failure occurs
+    """
+    if settings.DEBUG or settings.TEST:
+        response_data = {"error": "Sorry, this feature is not available in development or test environment."}
+        return Response(response_data)
+
+    # Check if the challenge phase exists or not
+    try:
+        challenge_phase = get_challenge_phase_model(challenge_phase_pk)
```
Comment (RishabhJain2018) [inline apps/jobs/views.py:2352]: There is no need to use try/except with the get_challenge_phase_model function as it automatically handles that case. Please see its definition.

File: apps/jobs/views.py [hunk 14]
```diff
@@ -2301,26 +2302,152 @@ def get_submission_file_presigned_url(request, challenge_phase_pk):
...
+
+        if not challenge.approved_by_admin:
+            response_data = {
+                "error": "Challenge is not yet approved by admin."
+            }
+            return Response(
+                response_data, status=status.HTTP_406_NOT_ACCEPTABLE
+            )
+
+        # if allowed email ids list exist, check if the user exist in that list or not
```
Comment (RishabhJain2018) [inline apps/jobs/views.py:2391]: We also have a blocked email Ids list and we should also check for that along with checking for allowed and blocked domains.

File: apps/jobs/views.py [hunk 15]
```diff
@@ -2301,26 +2302,152 @@ def get_submission_file_presigned_url(request, challenge_phase_pk):
...
+        submission = get_submission_model(submission_pk)
+        file_key_on_s3 = "{}/{}".format(
+            settings.MEDIAFILES_LOCATION, submission.input_file.name
+        )
+        response = complete_s3_multipart_file_upload(
+            file_parts, upload_id, file_key_on_s3, challenge.pk
+        )
+        if response.get("error"):
+            response_data = response
+            return Response(response_data, status=status.HTTP_400_BAD_REQUEST)
```
Comment (RishabhJain2018) [inline apps/jobs/views.py:2440]: Same for the `return` statement above.

File: apps/challenges/urls.py [hunk 16]
```diff
@@ -201,4 +201,9 @@
...
+        r"^phases/(?P<challenge_phase_pk>[0-9]+)/annotation_file_upload_complete/$",
```
Comment (RishabhJain2018) [inline apps/challenges/urls.py:205]: `annotation_file_upload_complete` -> `finish_annotation_file_upload`

File: apps/jobs/urls.py [hunk 17]
```diff
@@ -99,5 +99,10 @@
         r"^phases/(?P<challenge_phase_pk>[0-9]+)/get_submission_file_presigned_url/$",
         views.get_submission_file_presigned_url,
         name="get_submission_file_presigned_url",
+    ),
+    url(
+        r"^phases/(?P<challenge_phase_pk>[0-9]+)/submission_file_upload_complete/(?P<submission_pk>[0-9]+)/$",
+        views.submission_file_upload_complete,
+        name="submission_file_upload_complete",
```
Comment (RishabhJain2018) [inline apps/jobs/urls.py:106]: `submission_file_upload_complete` -> `finish_submission_file_upload`

**Review by RishabhJain2018 — CHANGES_REQUESTED** [review]
@Ram81 Can we please add the tests for new API's.

File: apps/jobs/views.py (hunk 14 above)
Comment (Ram81) [reply apps/jobs/views.py:2394]: Check here https://github.com/Cloud-CV/EvalAI/pull/3052/files#diff-f56c8d0075d862c83b8c37b0e742aec33542a7f1e9681b2cebdfc6306e8ee920R2411 it is already added

**Comment by Ram81** [comment]
@RishabhJain2018 this is also ready to be merged

File: apps/challenges/views.py (hunk 6 above)
Comment (RishabhJain2018) [reply apps/challenges/views.py:2694]: @Ram81 Please resolve this one as well.

File: apps/jobs/views.py (hunk 13 above)
Comment (RishabhJain2018) [reply apps/jobs/views.py:2355]: @Ram81 Please resolve this as well.

**Review by RishabhJain2018 — CHANGES_REQUESTED** [review]
@Ram81 We should also add tests for the newly added APIs.

**Comment by Ram81** [comment]
@RishabhJain2018 I've fixed the issues mentioned in review comments. I'll add the tests as a separate PR.

File: apps/jobs/views.py [hunk 18]
```diff
@@ -2301,26 +2302,158 @@ def get_submission_file_presigned_url(request, challenge_phase_pk):
...
+        submission_pk {int} -- Submission primary key
+    Returns:
+         Response Object -- An object containing the presignd url and submission id, or an error message if some failure occurs
+    """
+    if settings.DEBUG or settings.TEST:
+        response_data = {"error": "Sorry, this feature is not available in development or test environment."}
+        return Response(response_data)
+
+    # Check if the challenge phase exists or not
+    try:
```
Comment (RishabhJain2018) [inline apps/jobs/views.py:2354]: @Ram81 Can you please remove try/except from here.

**Review by RishabhJain2018 — CHANGES_REQUESTED** [review]

**Review by RishabhJain2018 — APPROVED** [review]

**Pull request merged**